src/stan/services/sample/hmc_nuts_dense_e_adapt.hpp

```diff
@@ -219,14 +349,15 @@ int hmc_nuts_dense_e_adapt(
     callbacks::interrupt& interrupt, callbacks::logger& logger,
     std::vector<InitWriter>& init_writer,
     std::vector<SampleWriter>& sample_writer,
-    std::vector<DiagnosticWriter>& diagnostic_writer) {
+    std::vector<DiagnosticWriter>& diagnostic_writer,
+    std::vector<callbacks::json_writer<Stream, Deleter>>& metric_writer) {
```

Comment:
Here we should just have the metric_writer itself be a template, similar to SampleWriter etc